### PR TITLE
Traverse User Profiles

### PR DESCRIPTION
**Pull request by bfollington**
- Allow dotted petnames
- Allow petname-only constructor for a Slashlink (i.e. @ben)
- Add docstring
- Tweak docstring
- Rework old test to assert new behaviour
- Track the path a user took to arrive at this profile
- Display full path to user profile in detail sheet
- Remove outdated init
- Update profile view to use new APIs
- Implement user profile actions + following checks

### CODE REVIEW

File: xcode/Subconscious/Shared/Components/Detail/UserProfileDetailView.swift [hunk 1]
```diff
@@ -112,15 +128,16 @@ struct UserProfileDetailModel: ModelProtocol {
...
             )
            
-        case .populate(let user, let statistics):
+        case .populate(let user, let spherePath, let statistics):
             var model = state
             model.user = user
             model.statistics = statistics
             model.recentEntries = (0...10).map { _ in EntryStub.dummyData(petname: user.petname) }
             model.topEntries = (0...10).map { _ in EntryStub.dummyData(petname: user.petname) }
             model.following = (1...10).map { _ in StoryUser.dummyData() }
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Detail/UserProfileDetailView.swift:137]: When we use real data we will need to fetch following status for everyone in the list. May be worth a global in-memory cache of Did -> following status, which gets invalidated / updated if you follow and unfollow.

File: xcode/Subconscious/Shared/Models/Petname.swift [hunk 2]
```diff
@@ -73,5 +73,10 @@ public struct Petname:
     public init?(formatting string: String) {
         self.init(Self.format(string))
     }
+    
+    public init?(petnames: [Petname]) {
+        let petnamePath = petnames.map({ s in s.verbatim }).joined(separator: ".")
+        self.init(petnamePath)
+    }
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Models/Petname.swift:80]: This will be removed / replaced by the code in https://github.com/subconsciousnetwork/subconscious/pull/489

**Comment by gordonbrander** [comment]
This is looking really good so far!

File: xcode/Subconscious/Shared/Components/AddressBook/FollowUserView.swift [hunk 3]
```diff
@@ -9,6 +9,59 @@ import SwiftUI
 import ObservableStore
 import CodeScanner
 
+struct FollowUserForm: View {
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/AddressBook/FollowUserView.swift:12]: I just extracted the form as seperate `View` for clarity, no changes

File: xcode/Subconscious/Shared/Components/Common/Profile/FollowUserSheet.swift [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+                text: Binding(
+                    get: { state.followUserForm.petname.value },
+                    send: send,
+                    tag: { input in
+                        .followUserForm(.petnameField(.setValue(input: input)))
+                    }
+                ),
+                caption: state.petnameFieldCaption ?? ""
+            )
+            .textFieldStyle(.roundedBorder)
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Common/Profile/FollowUserSheet.swift:180]: This needs to handle validation errors

File: xcode/Subconscious/Shared/Components/Common/Profile/UserProfileView.swift [hunk 5]
```diff
@@ -108,33 +114,139 @@ struct UserProfileView: View {
...
+    }
+    
+    func follow(state: UserProfileDetailModel, send: @escaping (UserProfileDetailAction) -> Void) -> some View {
+      self.modifier(FollowModifier(state: state, send: send))
+    }
+    
+    func metaSheet(state: UserProfileDetailModel, send: @escaping (UserProfileDetailAction) -> Void) -> some View {
+      self.modifier(MetaSheetModifier(state: state, send: send))
+    }
+}
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Common/Profile/UserProfileView.swift:135]: I had too many `.sheet`, `.alert` and `.confirmationDialog`s chained on this view for the typechecker, so I broke it into subcomponents via `ViewModifier`s.

File: xcode/Subconscious/Shared/Components/Detail/UserProfileDetailMetaSheet.swift [hunk 6]
```diff
@@ -197,7 +197,7 @@ struct UserProfileDetailMetaSheet: View {
             }
         }
         .presentationDragIndicator(.hidden)
-        .presentationDetents([.medium, .large])
+        .presentationDetents([.medium])
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Detail/UserProfileDetailMetaSheet.swift:200]: No reason to fill the screen given how limited the options are in the Profile Metasheet

File: xcode/Subconscious/Shared/Services/AddressBookService.swift [hunk 7]
```diff
@@ -75,27 +88,85 @@ class AddressBookService {
...
+    }
+    
+    func followUser(did: Did, petname: Petname, preventOverwrite: Bool = false) throws {
         if try self.noosphere.identity() == did.id {
             throw AddressBookError.cannotFollowYourself
         }
         
+        if try hasEntryForPetname(petname: petname) && preventOverwrite {
+            throw AddressBookError.invalidAttemptToOverwitePetname
+        }
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Services/AddressBookService.swift:151]: It seems like you should be able to put `preventOverwrite` first here to shortcircuit the logic, but Swift scolds me and says `try` cannot be next to `&&` (even with bracketing).

I chose this && format over a nested if statement, but it has _slightly_ worse performance characteristics like this.

File: xcode/Subconscious/Shared/Components/Detail/UserProfileDetailView.swift [hunk 8]
```diff
@@ -131,6 +194,121 @@ struct UserProfileDetailModel: ModelProtocol {
...
+            guard let did = state.followUserSheet.followUserForm.did.validated else {
+                return Update(state: state)
+            }
+            guard let petname = state.followUserSheet.followUserForm.petname.validated else {
+                return Update(state: state)
+            }
+            
+            let fx: Fx<UserProfileDetailAction> =
+                environment.data.addressBook
+                .followUserAsync(did: did, petname: petname, preventOverwrite: true)
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Detail/UserProfileDetailView.swift:234]: `preventOverwrite` will throw if there's a collision, which is intended here.

File: xcode/Subconscious/Shared/Components/Detail/UserProfileDetailView.swift [hunk 9]
```diff
@@ -131,6 +194,121 @@ struct UserProfileDetailModel: ModelProtocol {
...
+            return update(state: state, action: .presentUnfollowConfirmation(true), environment: environment)
+            
+        case .attemptUnfollow:
+            guard let petname = state.user?.petname, let did = state.user?.did else {
+                return Update(state: state)
+            }
+            
+            let fx: Fx<UserProfileDetailAction> =
+            environment.data.addressBook
+                .unfollowUserAsync(did: did)
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Detail/UserProfileDetailView.swift:285]: We unfollow by _DID_ here because the petname may not match what's in your address book.

File: xcode/Subconscious/Shared/Components/Form/FollowUserForm.swift [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+    static func validatePetname(petname: String) -> Petname? {
+        Petname(petname)
+    }
+    
+    static func update(state: Self, action: Action, environment: Environment) -> Update<Self> {
+        switch action {
+        case .didField(let action):
+            return DidFieldCursor.update(state: state, action: action, environment: FormFieldEnvironment())
+        case .petnameField(let action):
+            return PetnameFieldCursor.update(state: state, action: action, environment: FormFieldEnvironment())
```
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Components/Form/FollowUserForm.swift:35]: Nit: wrap line. https://google.github.io/swift/#function-calls

File: xcode/Subconscious/Shared/Components/Form/FollowUserForm.swift [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+        }
+    }
+}
+
+struct PetnameFieldCursor: CursorProtocol {
+    static func get(state: FollowUserFormModel) -> FormField<String, Petname> {
+        state.petname
+    }
+    
+    static func set(state: FollowUserFormModel, inner: FormField<String, Petname>) -> FollowUserFormModel {
```
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Components/Form/FollowUserForm.swift:45]: Nit: wrap line. https://google.github.io/swift/#function-calls

File: xcode/Subconscious/Shared/Components/Form/FollowUserForm.swift [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+        FollowUserFormAction.petnameField(action)
+    }
+}
+
+struct DidFieldCursor: CursorProtocol {
+    static func get(state: FollowUserFormModel) -> FormField<String, Did> {
+        state.did
+    }
+    
+    static func set(state: FollowUserFormModel, inner: FormField<String, Did>) -> FollowUserFormModel {
```
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Components/Form/FollowUserForm.swift:61]: Nit: wrap line https://google.github.io/swift/#function-calls

File: xcode/Subconscious/Shared/Models/Petname.swift [hunk 13]
```diff
@@ -17,6 +17,7 @@ public struct Petname:
     LosslessStringConvertible
 {
     private static let petnameRegex = /([\w\d\-]+)(\.[\w\d\-]+)*/
+    private static let petnameSuffixRegex = /^(?<petname>(.*?))(?<separator>-+)?(?<suffix>(\d+))?$/
```
[the remark this answers is not in the document]
Comment (gordonbrander) [reply xcode/Subconscious/Shared/Models/Petname.swift:20]: Could we simplify this by saying the separator is always one dash?

File: xcode/Subconscious/Shared/Models/Petname.swift [hunk 14]
```diff
@@ -80,5 +81,22 @@ public struct Petname:
...
+    /// A plain petname e.g. `ziggy` becomes `ziggy-1`
+    /// But `ziggy-1` becomes `ziggy-2` etc.
+    public func increment() -> Petname? {
+        guard let match = description.wholeMatch(of: Self.petnameSuffixRegex),
+              let separator = match.output.separator
+        else {
+            return Petname(formatting: verbatim + "-1")
+        }
+        
+        if let numberString = match.output.suffix, let number = Int(numberString) {
```
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Models/Petname.swift:95]: Nit: wrap lets

File: xcode/Subconscious/Shared/Services/AddressBookService.swift [hunk 15]
```diff
@@ -34,77 +44,201 @@ class AddressBookService {
     private(set) var database: DatabaseService
     private var addressBook: [AddressBookEntry]?
     
+    private static let MAX_ATTEMPTS_TO_INCREMENT_PETNAME = 99
```
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Services/AddressBookService.swift:47]: 👍

File: xcode/Subconscious/Shared/Services/AddressBookService.swift [hunk 16]
```diff
@@ -34,77 +44,201 @@ class AddressBookService {
     private(set) var database: DatabaseService
```
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Services/AddressBookService.swift:44]: Our new appraoch is to make all dependency services private.

File: xcode/Subconscious/Shared/Services/AddressBookService.swift [hunk 17]
```diff
@@ -34,77 +44,201 @@ class AddressBookService {
...
     
+    private func invalidateCache() {
+        addressBook = nil
+    }
+    
     /// Get the full list of entries in the address book.
     /// This is cached after the first use unless requested using refetch.
     /// If an error occurs producing the entries the resulting list will be empty.
-    func listEntries(refetch: Bool = false) -> AnyPublisher<[AddressBookEntry], Never> {
+    func listEntriesAsync(refetch: Bool = false) -> AnyPublisher<[AddressBookEntry], Error> {
```
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Services/AddressBookService.swift:67]: As discussed #501 introduces a conflicting change and we probably want to adopt the approach taken in that PR.

**Review by gordonbrander — CHANGES_REQUESTED** [review]

File: xcode/Subconscious/Shared/Models/Petname.swift (hunk 13 above)
[the remark this answers is not in the document]
Comment (bfollington) [reply xcode/Subconscious/Shared/Models/Petname.swift]: I played with this but if you look at the test cases this was the best way I could find to produce the "correct" petname in all cases.

The corner cases I'm was covering with this are things like `ben-----`  which should become `ben-----1` and `sarah99` which should become `sarah99-1` _not_ `sarah100` etc.

[the remark this answers is not in the document]
Comment (gordonbrander) [reply xcode/Subconscious/Shared/Models/Petname.swift]: 👍  that make sense. To clarify, I meant it could be possible to simplify the regex to match a single dash, rather than all. E.g.

Current Regex (matches many dashes):

Screen Shot 2023-04-18 at 8 55 05 AM

Possible alternative (matches just the one dash related to suffix):

Screen Shot 2023-04-18 at 8 54 51 AM

These are functionally equivalent for our purposes, but the second seemed closer to the intended logic to my eye.

Caveat: we are deep into nit territory, and no change or action is needed.

**Review by gordonbrander — APPROVED** [review]
Approve
